x-pack/plugins/apm/public/components/shared/DiscoverButtons/DiscoverButton.tsx

```diff
@@ -6,34 +6,44 @@
 
 import React from 'react';
 import { StringMap } from 'x-pack/plugins/apm/typings/common';
-import {
-  getAPMIndexPattern,
-  ISavedObject
-} from '../../../services/rest/savedObjects';
+import { getAPMIndexPattern } from '../../../services/rest/savedObjects';
 import { KibanaLink } from '../../../utils/url';
 
 interface Props {
   query: StringMap;
 }
 
 interface State {
-  indexPattern?: ISavedObject;
+  query: StringMap;
+}
+
+export async function setApmIndexPatternQuery(query: StringMap) {
+  if (query._a && query._a.index) {
+    return query;
+  }
+
+  const indexPattern = await getAPMIndexPattern();
+  const id = indexPattern && indexPattern.id;
+
+  if (!query._a) {
+    query._a = {};
+  }
+
+  query._a.index = id;
+
+  return query;
```

Comment:
Mutating `query` might give some surprises down the road. Can you make it immutable, and rename to `getQueryWithApmIndexPattern `?